## ML ZOOMCAMP HOMEWORK 1

In [1]:
# import all necessary libraries
import numpy as np
import pandas as pd

#### Question 1
     What's the version of Pandas that you installed?

In [2]:
pd.__version__

'1.5.3'

In [3]:
# load dataset
url = 'https://raw.githubusercontent.com/alexeygrigorev/datasets/master/laptops.csv'
laptopdata = pd.read_csv(url)

In [4]:
# Check first 5 and last 5 rows of dataframe
laptopdata

Laptop       Status   Brand  \
0     ASUS ExpertBook B1 B1502CBA-EJ0436X Intel Core...          New    Asus   
1     Alurin Go Start Intel Celeron N4020/8GB/256GB ...          New  Alurin   
2     ASUS ExpertBook B1 B1502CBA-EJ0424X Intel Core...          New    Asus   
3     MSI Katana GF66 12UC-082XES Intel Core i7-1270...          New     MSI   
4     HP 15S-FQ5085NS Intel Core i5-1235U/16GB/512GB...          New      HP   
...                                                 ...          ...     ...   
2155  Razer Blade 17 FHD 360Hz Intel Core i7-11800H/...  Refurbished   Razer   
2156  Razer Blade 17 FHD 360Hz Intel Core i7-11800H/...  Refurbished   Razer   
2157  Razer Blade 17 FHD 360Hz Intel Core i7-11800H/...  Refurbished   Razer   
2158  Razer Book 13 Intel Evo Core i7-1165G7/16GB/1T...  Refurbished   Razer   
2159  Razer Book FHD+ Intel Evo Core i7-1165G7/16GB/...  Refurbished   Razer   

           Model                CPU  RAM  Storage Storage type       GPU  \
0     ExpertBook      Intel Core i5    8      512          SSD       NaN   
1             Go      Intel Celeron    8      256          SSD       NaN   
2     ExpertBook      Intel Core i3    8      256          SSD       NaN   
3         Katana      Intel Core i7   16     1000          SSD  RTX 3050   
4            15S      Intel Core i5   16      512          SSD       NaN   
...          ...                ...  ...      ...          ...       ...   
2155       Blade      Intel Core i7   16     1000          SSD  RTX 3060   
2156       Blade      Intel Core i7   16     1000          SSD  RTX 3070   
2157       Blade      Intel Core i7   32     1000          SSD  RTX 3080   
2158        Book  Intel Evo Core i7   16     1000          SSD       NaN   
2159        Book  Intel Evo Core i7   16      256          SSD       NaN   

      Screen Touch  Final Price  
0       15.6    No      1009.00  
1       15.6    No       299.00  
2       15.6    No       789.00  
3       15.6    No      1199.00  
4       15.6    No       669.01  
...      ...   ...          ...  
2155    17.3    No      2699.99  
2156    17.3    No      2899.99  
2157    17.3    No      3399.99  
2158    13.4   Yes      1899.99  
2159    13.4   Yes      1699.99  

[2160 rows x 12 columns]

#### Question 2
    How many records are in the dataset?

In [5]:
laptopdata.shape

(2160, 12)

There are 2160 rows and 12 columns. This implies there are 2160 records entered in the dataset.

#### Question 3
    How many laptop brands are presented in the dataset?

In [6]:
# Count the number of unique laptop brands
laptop_brands = laptopdata['Brand'].nunique()
laptop_brands

27

There are 27 laptop brands present in the dataset.

#### Question 4

    How many columns in the dataset have missing values?

In [7]:
laptopdata.isnull().sum()

Laptop             0
Status             0
Brand              0
Model              0
CPU                0
RAM                0
Storage            0
Storage type      42
GPU             1371
Screen             4
Touch              0
Final Price        0
dtype: int64

Columns with mssing data are storage type and GPU.

#### Question 5
    What's the maximum final price of Dell notebooks in the dataset?

In [8]:
# Filter the dataset for Dell notebooks and find the maximum final price
max_dell_price = laptopdata[laptopdata['Brand'] == 'Dell']['Final Price'].max()
max_dell_price

3936.0

The maximum final price of Dell notebooks is 3936.0

#### Question 6
    Median value of Screen
    a. Find the median value of Screen column in the dataset.
    b. Next, calculate the most frequent value of the same Screen column.
    c. Use fillna method to fill the missing values in Screen column with the most frequent value from the previous step.
    d. Now, calculate the median value of Screen once again.

In [9]:
# Calculate the median value of the Screen column
initial_median_screen = laptopdata['Screen'].median()

In [10]:
initial_median_screen

15.6

In [12]:
# Calculate the most frequent value of the Screen column
most_frequent_screen = laptopdata['Screen'].mode()[0]

In [13]:
most_frequent_screen

15.6

In [14]:
# Use fillna method to fill missing values in Screen column with the most frequent value
laptopdata['Screen'].fillna(most_frequent_screen, inplace=True)

In [15]:
# Recalculate the median value of Screen
final_median_screen = laptopdata['Screen'].median()

In [16]:
final_median_screen

15.6

#### Question 7
    Sum of weights
    a. Select all the "Innjoo" laptops from the dataset.
    b. Select only columns RAM, Storage, Screen.
    c. Get the underlying NumPy array. Let's call it X.
    d. Compute matrix-matrix multiplication between the transpose of X and X. To get the transpose, use X.T. Let's call the result XTX.
    e. Compute the inverse of XTX.
    f. Create an array y with values [1100, 1300, 800, 900, 1000, 1100].
    g. Multiply the inverse of XTX with the transpose of X, and then multiply the result by y. Call the result w.
    h. What's the sum of all the elements of the result?

In [17]:
# Select all "Innjoo" laptops and relevant columns (RAM, Storage, Screen)
innjoo_laptops = laptopdata[laptopdata['Brand'] == 'Innjoo'][['RAM', 'Storage', 'Screen']]

In [18]:
# Get the underlying NumPy array (X)
X = innjoo_laptops.to_numpy()

In [19]:
# Compute XTX (matrix-matrix multiplication between transpose of X and X)
XTX = X.T @ X

In [20]:
# Compute the inverse of XTX
XTX_inv = np.linalg.inv(XTX)

In [21]:
# Create array y
y = np.array([1100, 1300, 800, 900, 1000, 1100])

In [22]:
# Compute the result w (XTX_inv * X.T * y)
w = XTX_inv @ X.T @ y

In [23]:
# Compute the sum of all elements of w
sum_of_w = np.sum(w)
sum_of_w

91.29988062995588